File: Java/020_ValidParentheses.java
package leetCode;

import java.util.Stack;

public class ValidParentheses {
	
	public boolean isValid(String s) {
        
		Stack<Character> stack = new Stack<Character>();
		
		for(int i = 0; i < s.length(); i++) {
			
			if(s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '[') {
				stack.push(s.charAt(i));
			} else {
				if(stack.isEmpty() || ((s.charAt(i)-stack.peek()) != 1 && (s.charAt(i)-stack.peek()) != 2)) {
					return false;
				} else  {
					stack.pop();
				}
			}
		}
		
		return stack.isEmpty();
    }
	
	public static void main(String[] args) {
		ValidParentheses v = new ValidParentheses();
		String s = ")[]";
		System.out.println(v.isValid(s));
		//System.out.println('}'-'{');
	}
}

File: Cpp/020_ValidParentheses.cpp
class Solution {
public:
	bool isValid(string s) {
		stack<char> stack;
		for (int i = 0; i < s.length(); i++) {
			if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
				stack.push(s[i]);
			} else {
				if (stack.empty() || ((s[i]-stack.top()) != 1 && (s[i]-stack.top()) != 2)) {
					return false;
				} else {
					stack.pop();
				}
			}
		}
		return stack.empty();
	}
};
